src/Features/Core/Portable/ConvertLinq/ConvertForEachToLinqQuery/AbstractConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,147 @@
+        /// </summary>
+        /// <param name="forEachStatement"></param>
+        /// <returns>
+        /// 1) extended nodes that can be converted into clauses
+        /// 2) identifiers introduced in nodes that can be converted
+        /// 3) statements that cannot be converted into clauses
+        /// 4) trailing comments to be added to the end
+        /// </returns>
+        protected abstract ForEachInfo<TForEachStatement, TStatement> CreateForEachInfo(
+            TForEachStatement forEachStatement, SemanticModel semanticModel);
+
+        /// <summary>
+        /// Tries to build a specific converter that covers e.g. Count, ToList, yield return or other similar cases.
+        /// </summary>
+        protected abstract bool TryBuildSpecificConverter(
+            ForEachInfo<TForEachStatement, TStatement> forEachInfo,
+            SemanticModel semanticModel,
+            TStatement statementCannotBeConverted,
+            CancellationToken cancellationToken,
+            out IConverter<TForEachStatement, TStatement> converter);
+
+        /// <summary>
+        /// Creates a default converter where foreach is joined with some children statements but other children statements are kept unmodified.
+        /// Example:
+        /// foreach(...)
+        /// {
+        ///     if (condition)
+        ///     {
+        ///        doSomething();
+        ///     }
+        /// }
+        /// is converted to
+        /// foreach(... where condition)
+        /// {
+        ///        doSomething(); 
+        /// }
+        /// </summary>
+        protected abstract IConverter<TForEachStatement, TStatement> CreateDefaultConverter(
+            ForEachInfo<TForEachStatement, TStatement> forEachInfo);
+
+        protected abstract SyntaxNode AddLinqUsing(
+            IConverter<TForEachStatement, TStatement> converter, SemanticModel semanticModel, SyntaxNode root);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var forEachStatement = root.FindNode(context.Span) as TForEachStatement;
+            if (forEachStatement == null)
+            {
+                return;
+            }
+
+            // Do not try to refactor queries with conditional compilation in them.
+            if (forEachStatement.ContainsDirectives)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (!TryBuildConverter(forEachStatement, semanticModel, cancellationToken, out var converter))
+            {
+                return;
+            }
+
+            if (semanticModel.GetDiagnostics(forEachStatement.Span, cancellationToken)
```

Comment:
>GetDiagnostics [](start = 30, length = 14)

Do we want to bail out only for syntax diagnostics or even for semantic diagnostics?

Reply:
I thought both. Do you think we should allow refactoring in case of semantic errors?

---
In reply to: [196953081](https://github.com/dotnet/roslyn/pull/27130#discussion_r196953081) [](ancestors = 196953081)

Reply:
I am fine with either, I was wondering whether you considered just bail on syntax first and hit some weird semantic error cases that caused you to exclude them as well.

---
In reply to: [196975855](https://github.com/dotnet/roslyn/pull/27130#discussion_r196975855) [](ancestors = 196975855,196953081)

Reply:
Cool. Thank you! Will check for syntax diagnostics in the beginning and for all in the end.

---
In reply to: [196976093](https://github.com/dotnet/roslyn/pull/27130#discussion_r196976093) [](ancestors = 196976093,196975855,196953081)